compiler/modules/front/const.bal

```diff
@@ -517,47 +520,84 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
+    else if expectedType == () || expectsFloat(tc, expectedType) {
+        result = floatFromFpLiteral(untypedLiteral);
+    }
+    else if expectsDecimal(tc, expectedType) {
+        result = check decimalFromFpLiteral(cx, untypedLiteral, startPos);
     }
     else {
-        return cx.semanticErr("invalid float literal", cause=result, pos=expr.startPos);
+        result = floatFromFpLiteral(untypedLiteral);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
 function foldIntLiteralExpr(FoldContext cx, t:SemType? expectedType, s:IntLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    float|int|error result;
-    string ty;
-    if expr.base == 10 && expectsFloat(expectedType) {
-        result = floatFromDecimalLiteral(expr.digits);
-        ty = "float"; 
-    }
-    else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
+    t:Context tc = cx.typeContext();
+    Position startPos = expr.startPos;
+    int|float|decimal result;
+    if expectedType == () || expectsInt(tc, expectedType) {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if expectsFloat(tc, expectedType) {
+        result = floatFromFpLiteral(expr.digits);
     }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+    else if expectsDecimal(tc, expectedType) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
+}
+
+function expectsInt(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:INT));
 }
 
-function expectsFloat(t:SemType? semType) returns boolean {
-    if semType == () {
-        return false;
+function expectsFloat(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:FLOAT));
+}
+
+function expectsDecimal(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:DECIMAL));
+}
+
+// Since the binary floating point literal is parsed correctly,
+// it is impossible to throw an error.
+function floatFromFpLiteral(string digits) returns float {
+    float|error f = float:fromString(digits);
+    if f is error {
+        panic err:impossible();
     }
-    else {
-        return t:isSubtypeSimple(t:intersect(semType, t:union(t:FLOAT, t:INT)), t:FLOAT);
+    return f;
+}
+
+// Even if the decimal floating point literal is parsed correctly,
+// overflow should throw an error.
+function decimalFromFpLiteral(FoldContext cx, string decimalStr, Position pos) returns decimal|FoldError {
+    decimal|error d = decimal:fromString(decimalStr);
+    if d is error {
+        return cx.semanticErr("invalid decimal literal", pos);
```

Comment:
Use spec consistent terminology: "invalid decimal floating point number".

Reply:
Also preserve the error from `decimal:fromString` in cause.

Reply:
Changed